### ISSUE TITLE
3.0.0-rc.1 is cannot be used because .smw.json cannot be created

### ISSUE THREAD

**Issue by Nikerabbit** [SemanticMediaWiki/SemanticMediaWiki#3506]
### Setup and configuration

- SMW version: 3.0.0-rc.1
- MW version: e53f26bde1d4a7763f243fa8a04f2a51131e995f (master)
- PHP version: 7.1.20-1+ubuntu16.04.1+deb.sury.org+1 (fpm-fcgi)
- DB system (MySQL, Blazegraph, etc.) and version: 10.0.36-MariaDB-0ubuntu0.16.04.1

### Issue
After running update.php and deploying new code, I get the message "missing upgrade key" as shown in https://www.semantic-mediawiki.org/wiki/Help:Upgrade

Output of update.php:
```
Setting upgrade key ...
   ... done.
```
But no file is created. This is because SMW assumes it has write permissions to the code directory, which is not the case when I run update.php. This information should be stored in the database itself or in a location that is writable.

### Steps to reproduce
Switch to different user which does not have a write permission to SMW directory before running update.php (or remove write permissions recursively from the directory)

**Comment by Nikerabbit** [comment]
Oh, and I suppose this can have poor interactions on wikifarms where multiple wikis are using shared codebase, so the data should be keyed by the dbname.

**Comment by Nikerabbit** [comment]
Btw, the update.php does not print any warning about failing to create the file. In PHP error logs I see
> PHP Warning:  file_put_contents(/srv/mediawiki/workdir/extensions/SemanticMediaWiki//.smw.json): failed to open stream: Permission denied in /srv/mediawiki/workdir/extensions/SemanticMediaWiki/src/Utils/File.php on line 23

PS. The exact slash is coming from https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/60c451a333a9c13fed93844053dd024bde53331e/src/SQLStore/Installer.php#L270 and can be removed.

**Comment by kghbln** [comment]
I see the issue so both doing it by dbname and to even be able to disable this for people who know what they are doing should be an easy way out.

**Comment by Nikerabbit** [comment]
> This information should be stored in the database itself

Actually, this might not be a good idea either. In my deployment script, the code and database updates are run off-line, followed by separate deployment step that updates symlinks. The live sites would go down between end of update.php and updating of the symlinks :(

**Comment by kghbln** [comment]
> This information should be stored in the database itself or in a location that is writable.

Probably the same location as used by MW core for [`$wgReadOnlyFile`](https://www.mediawiki.org/wiki/Manual:$wgReadOnlyFile) a functionality used heavily and successfully by some wikis I take care of.

**Comment by kghbln** [comment]
This may be a blocking issue for the planned release of SMW 3.0.0. To overcome this and to offer more flexibility a configuration parameter, e.g. `$smwgConfigFileDir` allowing to specify a writable directory appears to be solution.

Let me come back to this.

**Comment by kghbln** [comment]
Thanks again for reporting this issue and sharing your findings. This is indeed a disagreeable but in the end I do not think that is should block the release for another fortnight or so. I will include a note in my release mail for people with similar circumstances.

**Comment by wolfgang42** [comment]
+1. If this file can't be written it should cause an error; I just spent a lot of time trying to figure out why I was getting the error "Semantic MediaWiki was installed and enabled but is missing an appropriate upgrade key. Please run MediaWiki's update.php or Semantic MediaWiki's setupStore.php maintenance script first." when I had already done so, before I figured out that the webserver didn't have permissions to write the file.

**Comment by mwjames** [comment]
> This information should be stored in the database itself

No, it shouldn't because we don't necessarily have access to an open DB connection. Furthermore, static file access is faster than initialize and reading a DB table/row at each request to compare a sha1 value.

> This is because SMW assumes it has write permissions to the code directory, which is not the case when

Considered in #3596 (introduce `smwgConfigFileDir` to specify a location that is writable and persitent)

> Btw, the update.php does not print any warning about failing to create the file. 

Considered in #3596 (in those cases it will now raise a `FileNotWritableException`)

>  wikifarms where multiple wikis are using shared codebase, so the data should be keyed by the dbname.

Considered in #3596
